# Variance Explained

The goal of this notebook is to add a variance explained calculated as
`chi2.cdf(pvalue, df=1)`


In [ ]:
import polars as pl
from scipy.stats import chi2


### Load the dataset from previous notebook

In [4]:
lead_variant_maf_dataset = pl.read_parquet("lead-maf-vep/*.parquet")


In [5]:
lead_variant_maf_dataset.shape


(2622098, 19)

In [4]:
lead_variant_maf_dataset.head()


shape: (5, 18)
┌───────────┬───────────┬───────────┬──────────┬───┬───────────┬───────────┬───────────┬───────────┐
│ studyId   ┆ studyLocu ┆ variantId ┆ beta     ┆ … ┆ majorPopu ┆ majorPopu ┆ majorPopu ┆ vepEffect │
│ ---       ┆ sId       ┆ ---       ┆ ---      ┆   ┆ lation    ┆ lationMAF ┆ lationAll ┆ ---       │
│ str       ┆ ---       ┆ str       ┆ f64      ┆   ┆ ---       ┆ ---       ┆ eleFreque ┆ struct[3] │
│           ┆ str       ┆           ┆          ┆   ┆ struct[2] ┆ f64       ┆ ncy       ┆           │
│           ┆           ┆           ┆          ┆   ┆           ┆           ┆ ---       ┆           │
│           ┆           ┆           ┆          ┆   ┆           ┆           ┆ list[stru ┆           │
│           ┆           ┆           ┆          ┆   ┆           ┆           ┆ ct[2]]    ┆           │
╞═══════════╪═══════════╪═══════════╪══════════╪═══╪═══════════╪═══════════╪═══════════╪═══════════╡
│ gtex_txre ┆ 9cf8bb458 ┆ 13_453426 ┆ -1.21864 ┆ … ┆ {"nfe",0. ┆ 0.000559  ┆ [{"nfe_ad ┆ {"intron_ │
│ v_prostat ┆ 3ab369310 ┆ 74_C_T    ┆          ┆   ┆ 0}        ┆           ┆ j",0.0005 ┆ variant", │
│ e_ensg000 ┆ 742a50dee ┆           ┆          ┆   ┆           ┆           ┆ 59}]      ┆ 0.1,null} │
│ 001…      ┆ 370…      ┆           ┆          ┆   ┆           ┆           ┆           ┆           │
│ schmiedel ┆ 998d6b254 ┆ 13_453428 ┆ 1.6927   ┆ … ┆ {"nfe",0. ┆ 0.112651  ┆ [{"nfe_ad ┆ {"intron_ │
│ _2018_tx_ ┆ b25f87d2b ┆ 25_C_T    ┆          ┆   ┆ 0}        ┆           ┆ j",0.1126 ┆ variant", │
│ th2_memor ┆ 500e587ad ┆           ┆          ┆   ┆           ┆           ┆ 51}]      ┆ 0.1,null} │
│ y_e…      ┆ 60b…      ┆           ┆          ┆   ┆           ┆           ┆           ┆           │
│ gtex_txre ┆ 4666d5e4f ┆ 13_453430 ┆ 0.935377 ┆ … ┆ {"nfe",0. ┆ 0.112192  ┆ [{"nfe_ad ┆ {"intron_ │
│ v_uterus_ ┆ b13431c59 ┆ 39_A_T    ┆          ┆   ┆ 0}        ┆           ┆ j",0.1121 ┆ variant", │
│ ensg00000 ┆ 37e457207 ┆           ┆          ┆   ┆           ┆           ┆ 92}]      ┆ 0.1,null} │
│ 170…      ┆ 30e…      ┆           ┆          ┆   ┆           ┆           ┆           ┆           │
│ gtex_txre ┆ 26633dba5 ┆ 13_453437 ┆ 0.521832 ┆ … ┆ {"nfe",0. ┆ 0.020081  ┆ [{"nfe_ad ┆ {"intron_ │
│ v_muscle_ ┆ 87823b0ad ┆ 65_C_T    ┆          ┆   ┆ 0}        ┆           ┆ j",0.0200 ┆ variant", │
│ ensg00000 ┆ d355f3e67 ┆           ┆          ┆   ┆           ┆           ┆ 81}]      ┆ 0.1,null} │
│ 170…      ┆ 881…      ┆           ┆          ┆   ┆           ┆           ┆           ┆           │
│ phlips_ex ┆ bece5b71e ┆ 13_453442 ┆ -0.67395 ┆ … ┆ {"nfe",0. ┆ 0.369299  ┆ [{"nfe_ad ┆ {"intron_ │
│ on_ipsc_e ┆ fb8e16685 ┆ 47_T_TTTT ┆          ┆   ┆ 0}        ┆           ┆ j",0.6307 ┆ variant", │
│ nsg000001 ┆ c1318cac9 ┆ TA        ┆          ┆   ┆           ┆           ┆ 01}]      ┆ 0.1,null} │
│ 709…      ┆ 864…      ┆           ┆          ┆   ┆           ┆           ┆           ┆           │
└───────────┴───────────┴───────────┴──────────┴───┴───────────┴───────────┴───────────┴───────────┘

In [6]:
# NOTE! Calculate variance explained requires removal of the studies that have nSamples = 0
df = (
    lead_variant_maf_dataset.filter(pl.col("nSamples").is_not_null())
    .select(
        "*",
        (pl.col("pValueMantissa") * 10).pow(pl.col("pValueExponent")).alias("pValue"),
        (pl.col("pValueMantissa") * 10)
        .pow(pl.col("pValueExponent").map_elements(lambda x: chi2.cdf(x, 1), pl.Float64()))
        .alias("cdf"),
    )
    .select(
        "*",
        (pl.col("cdf") / pl.col("nSamples")).alias("varianceExplained"),
    )
)


In [ ]:
df.write_parquet("lead-maf-vep-ve.parquet")


In [ ]:
df.shape[0] - lead_variant_maf_dataset.shape[0]


-1050

Exactly 1050 samples have no `nSamples` disallowing us to calculate the varianceExplained